packages/tabsheet/src/vaadin-tabsheet.js

```diff
@@ -3,7 +3,269 @@
+      /**
+       * The tabsheet's orientation. Possible values are `horizontal|vertical`
+       * @type {!TabSheetOrientation}
+       */
+      orientation: {
+        reflectToAttribute: true,
+        value: 'horizontal',
+        type: String,
+        observer: '__orientationChanged',
+      },
+
+      /**
+       * The list of `<vaadin-tab>`s from which a selection can be made.
+       * It is populated from the elements passed inside the slotted
+       * `<vaadin-tabs>`, and updated dynamically when adding or removing items.
+       *
+       * Note: unlike `<vaadin-combo-box>`, this property is read-only.
+       * @type {!Array<!Tab> | undefined}
+       */
+      items: {
+        type: Array,
+        readOnly: true,
+        notify: true,
+      },
+
+      /**
+       * The index of the selected tab.
+       */
+      selected: {
+        value: 0,
+        type: Number,
+        reflectToAttribute: true,
+        notify: true,
+      },
+
+      /**
+       * The slotted <vaadin-tabs> element.
+       */
+      __tabs: {
+        type: Object,
+      },
+
+      /**
+       * The panel elements.
+       */
+      __panels: {
+        type: Array,
+      },
+    };
+  }
+
+  /** @protected */
+  ready() {
+    super.ready();
+
+    this.role = 'tablist';
+
+    const tabsItemsChangedListener = () => this._setItems(this.__tabs.items);
+
+    const tabsSelectedChangedListener = () => {
+      this.selected = this.__tabs.selected;
+    };
+
+    // Observe the tabs slot for a <vaadin-tabs> element.
+    this.addController(
+      new (class extends SlotController {
+        constructor(host) {
+          super(host, 'tabs');
+        }
+
+        initCustomNode(tabs) {
+          if (!(tabs instanceof Tabs)) {
+            throw Error('The "tabs" slot of a <vaadin-tabsheet> must only contain a <vaadin-tabs> element!');
+          }
+          tabs.addEventListener('items-changed', tabsItemsChangedListener);
+          tabs.addEventListener('selected-changed', tabsSelectedChangedListener);
+          this.host.__tabs = tabs;
+        }
+
+        teardownNode(tabs) {
+          tabs.removeEventListener('items-changed', tabsItemsChangedListener);
+          tabs.removeEventListener('selected-changed', tabsSelectedChangedListener);
+          this.host._setItems([]);
+        }
+      })(this),
+    );
+
+    // Observe the panels slot for nodes. Set the assigned element nodes as the __panels array.
+    const panelSlot = this.shadowRoot.querySelector('#panel-slot');
+    this.__panelsObserver = new FlattenedNodesObserver(panelSlot, () => {
+      this.__panels = Array.from(panelSlot.assignedNodes({ flatten: true })).filter(
+        (node) => node.nodeType === Node.ELEMENT_NODE,
+      );
+    });
+  }
+
+  static get observers() {
+    return [
+      '__itemsOrPanelsChanged(items, __panels)',
+      '__selectedTabItemChanged(selected, items, __panels)',
+      '__propagateTabsProperties(__tabs, orientation, selected)',
```

Comment:
We could simplify properties propagation using `DelegateStateMixin` e.g. like this:

```js
 static get delegateProps() {
  return ['orientation', 'selected'];
}
```

Then we just need to use `this.setTarget = tabs` when assigning `__tabs` element.

Reply:
Done